[13기 박규태] 스텝2 상태 관리로 메뉴 관리하기

gtparkk opened the pull request:
- [x] [localStorage](https://developer.mozilla.org/ko/docs/Web/API/Window/localStorage)에 데이터를 저장하여 새로고침해도 데이터가 남아있게 한다.
- [x]  에스프레소, 프라푸치노, 블렌디드, 티바나, 디저트 각각의 종류별로 메뉴판을 관리할 수 있게 만든다.
  - [x]  페이지에 최초로 접근할 때는 에스프레소 메뉴가 먼저 보이게 한다.
- [x]  품절 상태인 경우를 보여줄 수 있게, 품절 버튼을 추가하고 `sold-out` class를 추가하여 상태를 변경한다.

gtparkk committed:
JS : step1 - addMenu

gtparkk committed:
JS : step1 - editMenu, deleteMenu

스텝1 미션 - 수정버튼, 삭제버튼 기능 구현

gtparkk committed:
JS : 2/13 코드리뷰 반영

gtparkk committed:
JS : 1차 코드리뷰 반영

gtparkk committed:
localStorage Read&Write

gtparkk committed:
MenuManaging  by Category

gtparkk committed:
Add sold-out function & Refactoring

gtparkk committed:
주석정리

gtparkk committed:
Merge branch 'ggooltae' of https://github.com/blackcoffee-study/moonbucks-menu into ggooltae

hustle-dev commented on src/js/index.js:
```diff
@@ -1,59 +1,131 @@
+
+const soldOutMenu = (e) => {
+    const menuId = e.target.parentNode.dataset.menuId;
+    menu[currentCategory][menuId].soldOut = !menu[currentCategory][menuId].soldOut; //undefined인 경우 false로 판단하므로 true로 변경됨.
```
`undefined`의 특성을 사용해서 이렇게 menu를 관리해도 좋지만, 명시적인 방법으로 메뉴를 관리하는 상태에 soldOut과 관련한 상태를 만들어 메뉴 이름과 관리해서 사용해 보는 것은 어떨까요?

gtparkk replied:
최초에 메뉴를 추가 했을때,  메뉴의 soldOut 상태를 false로 저장하는 방법 말씀이신가요? 괜찮은 아이디어 같습니다!

관련해서 궁금한사항은, undefined의 특성을 사용해서 menu를 관리했을때에 비해  soldOut관련한 상태를 명시적으로 만든경우,  어떤 장점이 있다고 생각하셔서 Comment 달아주신건지 질문드립니다!

hustle-dev replied:
제가 생각하기에 `undefined`를 사용시 협업을 하는 과정에서 다른 개발자가 
이 코드를 이해하기 어려울 수도 있겠다 라는 생각이 들었어요.

이 코드를 작성한 이유를 아마 궁금해 할것 같다는 생각이 들었고, 혹여라도 만약 자바스크립트에 대해
잘 알지 못하는 사람이 코드를 보았을 때, 혼동을 줄 수 있겠다 라는 생각이 들어서 객체 안에서 soldOut 상태를 관리한다면
명시적으로 어떤식으로 동작하는 지 알기 편할것 같다는 생각을 하였습니다.

hustle-dev commented on src/js/index.js:
```diff
@@ -1,59 +1,131 @@
+    });
+
+    // 메뉴리스트에 클릭 이벤트 추가 (이벤트위임(Delegation) (수정,삭제,품절 버튼))
+    $('#menu-list').addEventListener('click', (e) => {
```
이 부분에서 수정, 삭제, 품절 버튼이외에 다른 부분을 클릭하면 이벤트가 발생하고 있으므로 
early return을 사용하면 좋을것 같아요

gtparkk replied:
아하, 그렇네요. early return 을 사용할 수 있는부분이 있는지, 나름 찾아봤는데 못봤던 부분이네요. 감사합니다 👍

hustle-dev commented on src/js/index.js:
```diff
@@ -1,59 +1,131 @@
+            removeMenu(e);
+            return;
+        }
+        if (e.target.classList.contains('menu-sold-out-button')) {
```
`e.target.classList.contains` 부분이 중복되어서 사용되고 있는데 함수로 빼는 방법도 고려하시면 좋을것 같습니다

gtparkk replied:
중복적으로 사용되니 함수화 하면 괜찮을것같네요! 아이디어 감사합니다 🙂

hustle-dev commented on src/js/store/index.js:
```diff
@@ -0,0 +1,10 @@
+    }
+}
+
+export default store;
```
스토어를 따로 분리해서 `localStorage`에 접근하는 함수를 빼와서 사용하는 부분은 좋은 부분인것 같아요. 배워갑니다 👍

Yeonlisa replied:
스토어 분리하여 모듈화하는게 좋은것같아요

hustle-dev commented on index.html:
```diff
@@ -52,33 +52,35 @@ <h1 class="text-center font-bold">🌝 문벅스 메뉴 관리</h1>
+              <ul id="menu-list" class="mt-3 pl-0"></ul>
             </div>
           </main>
         </div>
```
메뉴가 확장됨에 따라 기존 id 명을 바꾸고 리팩토링하신것 같아 좋아보이네요 👍
저도 반영하겠습니다.

hustle-dev commented on src/js/index.js:
```diff
@@ -1,59 +1,131 @@
+
+//페이지 처음 진입시 호출하는 함수
+const init = () => {
+    if (store.getLocalStorage()) {
```
이 부분을 단축 논리 평가 계산을 통해 코드 작성하면 더 깔끔해질 것 같아 한번 고려해보시면 좋을 것 같아요!

gtparkk replied:
단축 논리평가계산은 처음 들어본 내용인데 한번 찾아보겠습니다. 감사합니다 !

Yeonlisa commented on src/js/index.js:
```diff
@@ -1,59 +1,131 @@
 //Enter키 입력시 새로고침 방지
-const espressoMenuForm = $('#espresso-menu-form');
-espressoMenuForm.setAttribute('onsubmit', "return false;");
+const menuForm = $('#menu-form');
+menuForm.setAttribute('onsubmit', "return false;");
```
변수명이 직관적이어서 좋고, setAttribute를 애용하시는것같아 배우고갑니다

Yeonlisa commented on src/js/index.js:
```diff
@@ -1,59 +1,131 @@
+        if (isCategoryButton) {
+            const categoryName = e.target.dataset.categoryName;
+            currentCategory = categoryName;
+            $('#category-title').textContent = `${e.target.textContent} 메뉴 관리`;
```
자꾸 textContent가 아닌 innerText를 쓰게되서 걱정이네요.. 그래도 깔끔해서 좋은것 같습니다.

FEsophie replied:
@Yeonlisa  연리사님 궁금한게 있는데용
혹시 textContent가 아닌 innerText를 쓸때 문제점이 있을까요??

Yeonlisa replied:
https://developer.mozilla.org/ko/docs/Web/API/Node/textContent 여기보고 알게되었습니다!

Yeonlisa commented on src/js/utils/dom.js:
```diff
@@ -0,0 +1,2 @@
+//DOM요소 선택 함수
+export const $ = (selector) => document.querySelector(selector);
```
DOM 역시 분리하여 모듈화하신게 인상적입니다

Yeonlisa reviewed (commented):
리뷰 마칩니다.

gtparkk committed:
[fix] 코드리뷰 피드백 반영

- 메뉴 클릭이벤트에서, 버튼을 클릭하지 않은경우 -> early return 반영
- e.target.classList.contains 중복사용으로 인한 함수화`checkClassList()`

FEsophie commented on src/js/index.js:
```diff
@@ -1,59 +1,134 @@
+    // 메뉴리스트에 클릭 이벤트 추가 (이벤트위임(Delegation) (수정,삭제,품절 버튼))
+    $('#menu-list').addEventListener('click', (e) => {
+        if (e.target.type != 'button') return; //버튼을 클릭하지 않은경우 빠른리턴
+        if (checkClassList(e.target, 'menu-edit-button')) {
```
저는 항상  matches 메서드로 엘리먼트를 체크를 해서  type 이라는 필드가 있는지 처음 알았네용 ㅎㅎ 👍
